### PR TITLE
Refactor the evaluator by removing handlers etc

### LINKED ISSUE #1313: The ensemble object should create the snapshot
Currently, the ensemble evaluator contains the logic of creating the snapshot. I think that logic should be moved inside the ensemble

### LINKED ISSUE #1690: Separate server logic and status logic in evaluator
Currently, the logic for handling connection, authorisation, etc and the logic for handling events and updating the logic is very interconnected. We should separate these and make the status updating logic into a separate module. 

In idea is that the status logic has an inbound queue, where events from dispatchers are put as soon as they are received and converted to CloudEvents. On the other side, there is an outbound queue where the Snapshots / SnapshotUpdates are put by the status updater, and then broadcasted to all clients via the websocket server.  

This module should have the same capabilities wrt. batching as the current implementation has.

### PR DESCRIPTION
**Pull request by frode-aarstad**
Resolves #1690
Resolves #1313
We have done the following:
- Moved the responsibility for updating the snapshot/tracker from evaluator to dispatcher (new name pending)
- Moved the snapshot creation code inside the ensemble

### CODE REVIEW

File: ert_shared/ensemble_evaluator/dispatch.py [hunk 1]
```diff
@@ -41,27 +48,86 @@ async def join(self):
...
+        await self._evaluator_callback(
+            identifiers.EVTYPE_EE_SNAPSHOT_UPDATE, snapshot_update_event
+        )
+
+    async def _ensemble_stopped_handler(self, events):
+        if self._ensemble.get_status() != ENSEMBLE_STATE_FAILED:
+            await self._evaluator_callback(
+                identifiers.EVTYPE_ENSEMBLE_STOPPED,
+                self._ensemble.update_snapshot(events),
+                events[0].data,  # TODO fix this
```
Comment (verveerpj) [inline ert_shared/ensemble_evaluator/dispatch.py:102]: Fix what?

File: ert_shared/ensemble_evaluator/dispatch.py [hunk 2]
```diff
@@ -41,27 +48,86 @@ async def join(self):
...
-                return function(*args, **kwargs)
-
-            return wrapper
-
-        return decorator
+        self._ensemble = ensemble
+        self._evaluator_callback = evaluator_callback
+        self._register_event_handlers()
+
+    def register_event_handler(self, event_types, function, batching=False):
```
Comment (verveerpj) [inline ert_shared/ensemble_evaluator/dispatch.py:58]: Are there cases were `batch=False` is used?

Comment (DanSava) [reply ert_shared/ensemble_evaluator/dispatch.py:58]: In the tests

Comment (verveerpj) [reply ert_shared/ensemble_evaluator/dispatch.py:58]: Minor thing, I guess, but if it is only used for testing, should we not be patching/mocking somehow instead?

**Review by verveerpj — APPROVED** [review]
LGTM.

File: ert_shared/ensemble_evaluator/dispatch.py (hunk 1 above)
Comment (DanSava) [reply ert_shared/ensemble_evaluator/dispatch.py:102]: Fix the fact that we are forced to get the element inside a list with only one element

Comment (DanSava) [reply ert_shared/ensemble_evaluator/dispatch.py:102]: And that still needs a fix :disappointed:

**Pull request merged**